### DISCUSSION TITLE
Issue passing bytes as stream to recognize song

### DISCUSSION THREAD

**Discussion by captain-cronch** [shazamio/ShazamIO#61] [Q&A] [answered]
Based on a closed issue, it looks like the ability to do this was added in version 0.0.6, but I am trying to figure out the proper way to get it to work.

I have tried using both pyaudio (which returns numerous ALSA lib and jack errors, but I am still able to record and save files) as well as alsaaudio to stream audio from my USB mic to shazamio.

The code I am using for pyaudio is as follows:

`import asyncio
import io
import pyaudio
from shazamio import Shazam

async def main():
    shazam = Shazam()
    
    chunk_size = 1024
    sample_format = pyaudio.paInt16
    channels = 1
    sample_rate = 44100
    record_seconds = 5
    
    p = pyaudio.PyAudio()
    stream = p.open(format=sample_format,
                    channels=channels,
                    rate=sample_rate,
                    frames_per_buffer=chunk_size,
                    input=True)
    
    audio_buffer = io.BytesIO()
    
    while True:
        data = stream.read(chunk_size)
        audio_buffer.write(data)
        
        if audio_buffer.tell() >= sample_rate * record_seconds * 2:
            audio_buffer.seek(0)
            out = await shazam.recognize_song(audio_buffer)
            print(out)
            audio_buffer = io.BytesIO()
    
    stream.stop_stream()
    stream.close()
    p.terminate()
    
loop = asyncio.get_event_loop()
loop.run_until_complete(main())
`

In both instances, I receive an error similar to this:

`Traceback (most recent call last):
  File "/home/user/shazamiotest2.py", line 37, in <module>
    loop.run_until_complete(main())
  File "/usr/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/home/user/shazamiotest2.py", line 30, in main
    out = await shazam.recognize_song(audio_buffer)
  File "/usr/local/lib/python3.9/dist-packages/shazamio/api.py", line 348, in recognize_song
    audio = self.normalize_audio_data(song)
  File "/usr/local/lib/python3.9/dist-packages/shazamio/converter.py", line 51, in normalize_audio_data
    audio = audio.set_sample_width(2)
AttributeError: 'NoneType' object has no attribute 'set_sample_width'`

Any help working out this issue would be greatly appreciated.
Thank you.

**Comment by dotX12** [comment]
@captain-cronch, hello, very busy week, I will try to reproduce the problem and see how to solve over the weekend!

**Reply by captain-cronch** [reply]
Thank you. For what it’s worth I’m using a raspberry pi 3 on the latest version of raspberry pi OS. 

I also tried the following and received the same ALSA lib errors, but the script took longer to finish than recording and saving an audio clip before passing it to Shazam. I didn’t let the script finish to see if I received a nontype error on sample width

```
import pyaudio
from shazamio import Shazam

shazam = Shazam()

CHUNK_SIZE = 1024
FORMAT = pyaudio.paInt16
CHANNELS = 1
RATE = 44100

p = pyaudio.PyAudio()

stream = p.open(format=FORMAT,
                channels=CHANNELS,
                rate=RATE,
                input=True,
                frames_per_buffer=CHUNK_SIZE)

audio_bytes = b''
while True:
    data = stream.read(CHUNK_SIZE)
    audio_bytes += data
    if len(audio_bytes) > 12000000:  # Stop after 12 seconds of audio
        break

song_info = shazam.recognize_song(audio_bytes)

print(song_info)

stream.stop_stream()
stream.close()
p.terminate()
```

**Comment by MoppyDev** [comment]
I actually got the same when trying to stream the data in initially. But I did find a way to workaround the issue in case someone runs into this.

```
import asyncio
from shazamio import Shazam
import json
import soundfile as sf
import sounddevice as sd
from io import BytesIO
from pydub import AudioSegment

RATE = 44100
CHUNK = 1024
SECONDS = 10

print("recording...")
data = sd.rec(int(SECONDS * RATE), samplerate=RATE, channels=1)
sd.wait()

file_format = "WAV"
memory_file = BytesIO( )
sf.write( memory_file, data, RATE, format = file_format )

memory_file.seek( 0 )

async def main(memory_file):
  shazam = Shazam()
  out = await shazam.recognize_song(AudioSegment.from_wav(memory_file))
  print(json.dumps(out, indent=10))

response = asyncio.run(main(memory_file))
```

Basically the workaround for me was to convert the bytes into a wav file in memory. I wanted this so to not keep creating audio files. Chunked this out from a bigger project I'm working on so excuse the code from looking pristine! 


Error without the AudioSegment.from_wav:

```
 normalize_audio_data
    audio = audio.set_sample_width(2)
AttributeError: 'NoneType' object has no attribute 'set_sample_width'
```

**Reply by MoppyDev** [reply, marked as answer]
Without AudioSegment you just need to getvalue on the bytes:

```
import asyncio
from shazamio import Shazam
import json
import soundfile as sf
import sounddevice as sd
from io import BytesIO

RATE = 44100
CHUNK = 1024
SECONDS = 10

print("recording...")
data = sd.rec(int(SECONDS * RATE), samplerate=RATE, channels=1)
sd.wait()

file_format = "WAV"
memory_file = BytesIO( )
sf.write( memory_file, data, RATE, format = file_format )

memory_file.seek( 0 )

async def main(memory_file):
  shazam = Shazam()
  out = await shazam.recognize_song(memory_file.getvalue())
  print(json.dumps(out, indent=10))

response = asyncio.run(main(memory_file))
```
